### ISSUE TITLE
In some cases num_positives = 0, how to ensure the num_positives >0

### ISSUE THREAD

**Issue by xingbod** [AdrianUng/keras-triplet-loss-mnist#2]
In some cases num_positives = 0, how to ensure the num_positives >0

**Comment by AdrianUng** [comment]
Hello, charlesLucky!

That's a good question. Could you please tell me what was the batch size that you used, which resulted in num_positives=0? This would happen if batch_size is too small.
In the initial FaceNet paper, the authors used a batch size of 1,800.

I hope that helps!

**Comment by xingbod** [comment]
My batch size is 8 only coz I don't have a big GPU memory, and my dataset is also a small dataset.

I try to avoid 0 cases by setting the num_positives with 1 if there are no positives:

`num_positives = math_ops.maximum(num_positives, 1)  # avoid nan`

I am not sure if this is good or not, this, of course, will introduce some error into the calculation.

Or maybe I need to schedule each batch for each iteration of the training?

**Comment by AdrianUng** [comment]
Yes! I would say that that is the biggest problem... try to run your experiment with a bigger GPU.
I am sure there is a way around that, specifically by using a data generator: [suggestion](https://github.com/noelcodella/tripletloss-keras-tensorflow/blob/8554a39add87488fb70a6a4b83bc049e11fe9cc7/tripletloss.py#L60)

You shouldn't give **num_positives** the value of 1 forcefully like that, because the problem lies on the previous line:
`num_positives = math_ops.reduce_sum(mask_positives)  `
which computes the sum of **mask_positives**, in your case being equal to 0.



How big is your dataset (I assume it's not MNIST)? also, what is the architecture you are considering?

**Comment by xingbod** [comment]
I am using inception-resnet, and the dataset only has 3000 around images. I will check the data generator to find possible solutions.

**Comment by AdrianUng** [comment]
You might want to consider smaller models for training:
- AlexNet, MobileNet
- VGG-S, M, F models from the Return of the Devil [paper](http://www.robots.ox.ac.uk/~vgg/publications/2014/Chatfield14/chatfield14.pdf), [weights](http://www.vlfeat.org/matconvnet/pretrained/)
- Smaller Resnets (like 8, 18, etc.), but the main problem here would be to find trained models on ImageNet...

I hope this helps!

**Comment by xingbod** [comment]
tks so much for all the help, I will look into those options
